File: ai-solutions/03/sonnet-4.5/js/solution.js
const fs = require('fs');

function findMaxNumber(digits, count) {
    const n = digits.length;
    if (count > n) return 0;
    
    // Greedy approach: select 'count' digits that form the largest number
    const result = [];
    let start = 0;
    
    for (let i = 0; i < count; i++) {
        // We need to pick 'count - i' more digits from the remaining string
        // So we can search up to index n - (count - i)
        const end = n - (count - i) + 1;
        
        // Find the maximum digit in the valid range
        let maxDigit = digits[start];
        let maxPos = start;
        
        for (let j = start; j < end; j++) {
            if (digits[j] > maxDigit) {
                maxDigit = digits[j];
                maxPos = j;
            }
        }
        
        result.push(maxDigit);
        start = maxPos + 1;
    }
    
    return parseInt(result.join(''));
}

function solve(input) {
    const lines = input.trim().split('\n');
    
    let part1 = 0;
    let part2 = 0;
    
    for (const line of lines) {
        const digits = line.trim();
        part1 += findMaxNumber(digits, 2);
        part2 += findMaxNumber(digits, 12);
    }
    
    console.log(part1);
    console.log(part2);
}

const input = fs.readFileSync('../../../../inputs/03.txt', 'utf8');
solve(input);

File: ai-solutions/03/sonnet-4.5/python/solution.py
def find_max_number(digits, count):
    n = len(digits)
    if count > n:
        return 0
    
    # Greedy approach: select 'count' digits that form the largest number
    result = []
    start = 0
    
    for i in range(count):
        # We need to pick 'count - i' more digits from the remaining string
        # So we can search up to index n - (count - i)
        end = n - (count - i) + 1
        
        # Find the maximum digit in the valid range
        max_digit = digits[start]
        max_pos = start
        
        for j in range(start, end):
            if digits[j] > max_digit:
                max_digit = digits[j]
                max_pos = j
        
        result.append(max_digit)
        start = max_pos + 1
    
    return int(''.join(result))

def solve(input_text):
    lines = input_text.strip().split('\n')
    
    part1 = 0
    part2 = 0
    
    for line in lines:
        digits = line.strip()
        part1 += find_max_number(digits, 2)
        part2 += find_max_number(digits, 12)
    
    print(part1)
    print(part2)

with open('../../../../inputs/03.txt', 'r') as f:
    input_text = f.read()
    solve(input_text)
